chain-signatures/contract/src/errors/mod.rs

```diff
@@ -0,0 +1,129 @@
+use std::borrow::Cow;
+mod impls;
+
+#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
+pub enum SignError {
+    #[error("Signature request has timed out.")]
+    Timeout,
+    #[error("Signature request has already been submitted. Please try again later.")]
+    PayloadCollision,
+    #[error("Malformed payload.")]
+    MalformedPayload,
+    #[error(
+        "This key version is not supported. Call latest_key_version() to get the latest supported version."
+    )]
+    UnsupportedKeyVersion,
+    #[error("Too many pending requests. Please try again later.")]
+    RequestLimitExceeded,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
+pub enum RespondError {
+    #[error("The provided signature is invalid.")]
+    InvalidSignature,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, thiserror::Error)]
+pub enum JoinError {
+    #[error("Account to join is already in the participant set.")]
+    JoinAlreadyParticipant,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, thiserror::Error)]
+pub enum PublicKeyError {
+    #[error("Derived key conversion failed.")]
+    DerivedKeyConversionFailed,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, thiserror::Error)]
+pub enum InitError {
+    #[error("Threshold cannot be greater than the number of candidates")]
+    ThresholdTooHigh,
+    #[error("Cannot load in contract due to missing state")]
+    ContractStateIsMissing,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, thiserror::Error)]
+pub enum VoteError {
+    #[error("Voting account is not in the participant set.")]
+    VoterNotParticipant,
+    #[error("Account to be kicked is not in the participant set.")]
+    KickNotParticipant,
+    #[error("Account to join is not in the candidate set.")]
+    JoinNotCandidate,
+    #[error("Number of participants cannot go below threshold.")]
+    ParticipantsBelowThreshold,
+    #[error("Update not found.")]
+    UpdateNotFound,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, thiserror::Error)]
+pub enum Common {
```

Comment:
Now that we have this pattern, we shouldn't be making a common kind. Instead we should separate out all of these into their own category such as:

```
enum ProtocolError {
    StateNotRunning,
    ...
}

enum InvalidState {
    RequestNotFound,
    
}

enum InvalidParameters {
    InsufficientDeposit,
    InsufficientGas,
    ...
}

```

Reply:
@ChaoticTempest This is opinionated and generates unnecessary code. Error representation for the user will not change, but for us, it can add some confusion. For example, `RequestNotFound` is `InvalidState` and `InvalidParameters` error at the same time. Same about `EpochMismatch`.

Reply:
hmm, I think `Common` is too ambiguous of name since that doesn't signify how that error could have happened. I rather us base it on opinionated items than make it ambiguous.

Then I guess `RequestNotFound` can go in `InvalidParameters` since I don't see it being using anywhere else besides for invalid parameters. `EpochMismatch` is more `InvalidState` though

Reply:
Done